pyartcd/pyartcd/jenkins.py

```diff
@@ -1,74 +1,147 @@
         params['TRIGGERED_FROM'] = build_url
+    return params
+
+
+def block_until_building(queue_item: QueueItem, delay: int = 5) -> str:
+    """
+    Watches a queue item and blocks until the scheduled build starts.
+
+    Returns the URL of the new build.
+    """
+
+    while True:
+        try:
+            data: dict = queue_item.poll()
+            build_number = data['executable']['number']
+            break
+        except (KeyError, TypeError):
+            logger.info('Build not started yet, sleeping for %s seconds...', delay)
+            time.sleep(delay)
+
+    logger.info('Build started: number = %s', build_number)
+    return f"{data['task']['url']}{build_number}"
+
+
+def start_build(job_name: str, params: dict, blocking: bool = False,
+                watch_building_delay: int = 5) -> Any[str, None]:
+    """
+    Starts a new Jenkins build
+
+    :param job_name: e.g. "aos-cd-builds/build%2Fbuild-sync"
+    :param params: a key-value collection to be passed to the build
+    :param blocking: if True, will block until the new builds completes; if False,
+                    start a new build in "fire-and-forget" fashion
+    :param watch_building_delay: Poll rate for building state
+
+    Returns the build result if blocking=True, None otherwise
+    """
 
-    # Build authorization header
-    auth = base64.b64encode(f'{service_account}:{token}'.encode()).decode()
-    auth_header = f'Basic {auth}'
-    headers = {'Authorization': auth_header}
-
-    # Build url
-    # If the job to be triggered is parametrized, use 'buildWithParameters' and send HTTP data
-    # Otherwise, just call the 'build' endpoint with empty data
-    url = f'https://buildvm.hosts.prod.psi.bos.redhat.com:8443/{job_path}'
-    if params:
-        url += '/buildWithParameters'
-    else:
-        url += '/build'
-
-    # Call API endpoint
-    logger.info('Triggering remote job /%s', job_path)
-    async with aiohttp.ClientSession(headers=headers) as session:
-        async with session.post(url, data=params) as response:
-            response.raise_for_status()
-            await response.text()
-
-
-async def trigger_ocp4(build_version: str):
-    await trigger_jenkins_job(
-        job_path='job/triggered-builds/job/ocp4',
-        params={'BUILD_VERSION': build_version}
+    logger.info('Starting new build for job: %s', job_name)
+    job = jenkins.get_job(job_name)
+    queue_item = job.invoke(build_params=params)
+    build_url = block_until_building(queue_item, watch_building_delay)
+    logger.info('Started new build at %s', build_url)
+
+    # If blocking==True, wait for the build to complete; get its status and return it
+    if blocking:
+        logger.info('Waiting for build to complete...')
+        build = Build(
+            url=build_url.replace(constants.JENKINS_UI_URL, constants.JENKINS_SERVER_URL),
+            buildno=int(build_url.split('/')[-1]),
+            job=job
+        )
+        build.block_until_complete()
```

Comment:
would you mind try
>>> b.job.jenkins.requester.post_and_confirm_status(f"{b.baseurl}/submitDescription",params={"Submit": "submit", "description":f"build by:{job.baseurl}"},data="",valid=[200])
to get the downstream job set in the description?
and can you add a wrapper to support starting a batch of builds or just return the build object? my use case looks like https://github.com/openshift-eng/aos-cd-jobs/pull/3563/files#diff-de01749278991bbde16a03048265c175ddc0aaf9a68f3f878396c59b821e4573R432-R433 I may need to invoke a list of jobs in a loop and monitoring the results in parallel

Reply:
Fixed the tracking part. Regarding the loop of parallel builds, let's address it in a new PR